--- a/cps/admin.py	2020-02-23 09:08:01.000000000 +0100
+++ b/cps/admin.py	2020-04-12 12:51:28.100811662 +0200
@@ -43,7 +43,7 @@
 from .web import admin_required, render_title_template, before_request, unconfigured, login_required_if_no_ano
 
 feature_support = {
-        'ldap': False, # bool(services.ldap),
+        'ldap': bool(services.ldap),
         'goodreads': bool(services.goodreads_support)
     }
 