--- a/cps/web.py	2021-01-17 16:28:28.000000000 +0100
+++ b/cps/web.py	2021-01-18 22:43:37.597311098 +0100
@@ -1394,7 +1394,14 @@
         if feature_support['oauth'] and (config.config_login_type == 2 or config.config_login_type == 3):
             logout_oauth_user()
     log.debug(u"User logged out")
-    return redirect(url_for('web.login'))
+#Début Modif Yunohost 1/1 - Redirect to Yunohost at exit
+#    return redirect(url_for('web.login'))
+    if config.config_login_type == constants.LOGIN_LDAP and services.ldap:
+        return redirect(request.host_url)
+    else:
+        return redirect(url_for('web.login'))
+#Fin Modif Yunohost 1/1
+ 
 
 
 
