--- ./web.py	2023-08-05 11:58:20.353503437 +0200
+++ ./web.py	2023-03-27 19:49:57.000000000 +0200
@@ -1411,11 +1411,7 @@
         if feature_support['oauth'] and (config.config_login_type == 2 or config.config_login_type == 3):
             logout_oauth_user()
     log.debug("User logged out")
-#    return redirect(url_for('web.login'))
-    if config.config_login_type == constants.LOGIN_LDAP:
-        return redirect(request.host_url + '/yunohost/sso/?action=logout')
-    else:
-        return redirect(url_for('web.login'))
+    return redirect(url_for('web.login'))
 
 
 # ################################### Users own configuration #########################################################
